client/src/main/java/org/eclipse/hono/client/impl/CommandTargetMapperImpl.java

```diff
@@ -0,0 +1,160 @@
+                                    } else {
+                                        targetInstanceObject = chooseTargetGatewayAndAdapterInstance(instancesArray);
+                                    }
+                                } catch (final ClassCastException e) {
+                                    LOG.error("Invalid result JSON", e);
+                                    return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+                                }
+                                final String targetDevice = targetInstanceObject.getString(DeviceConnectionConstants.FIELD_PAYLOAD_DEVICE_ID);
+                                final String targetAdapterInstance = targetInstanceObject.getString(DeviceConnectionConstants.FIELD_ADAPTER_INSTANCE_ID);
+                                if (targetDevice == null || targetAdapterInstance == null) {
+                                    LOG.error("Invalid result JSON, missing target device and/or adapter instance");
+                                    return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+                                }
+                                if (!targetDevice.equals(deviceId)) {
+                                    // target device is a gateway
+                                    if (!viaGateways.contains(targetDevice)) {
+                                        LOG.error("Invalid result JSON, target gateway {} is not in via gateways list", targetDevice);
+                                        return Future.failedFuture(new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR));
+                                    }
+                                    span.setTag(MessageHelper.APP_PROPERTY_GATEWAY_ID, targetDevice);
+                                }
+                                span.finish();
+                                return Future.succeededFuture(targetInstanceObject);
+                            });
+                }).recover(t -> {
+                    LOG.debug("Error getting target gateway and adapter instance", t);
+                    TracingHelper.logError(span, t);
+                    Tags.HTTP_STATUS.set(span, ServiceInvocationException.extractStatusCode(t));
+                    span.finish();
+                    return Future.failedFuture(t);
+                });
+    }
+
+    /**
+     * Chooses the target gateway and adapter instance from the given list of entries.
+     * <p>
+     * This method returns first entry from the given list.
+     * <p>
+     * Subclasses may override this method in order to apply a different algorithm.
+     *
+     * @param instancesArray The JSON array containing the target gateway and adapter instance entries to choose from.
+     * @return The chosen JSON object.
+     */
+    protected JsonObject chooseTargetGatewayAndAdapterInstance(final JsonArray instancesArray) {
```

Comment:
For such a mechanism, the `CommandTargetMapper.getTargetGatewayAndAdapterInstance` method would have to be changed so that it can return a result JSON array.

The implementation of that mechanism would then be where that method is called, ie. in `MappingAndDelegatingCommandHandler.mapAndDelegateIncomingCommand`.

I had decided against such a mechanism mainly because I think it's unclear how a gateway could either say "I as gateway reject the command because I can't handle it" (ie. "try next gateway") or "I reject the command because the target device has rejected it" (ie. "no use in trying other gateways").
One could argue that this is not important, that it wouldn't matter to try the other gateways. But I don't really like that approach, causing traffic and a delay in acknowledging the command for nothing in many cases.

I had rather thought of making it possible to override the `chooseTargetGatewayAndAdapterInstance` method and let a specific implementation better choose the one gateway that the command will be sent to, e.g. by taking extra properties returned in the device connection result into account or by querying the registration service for some gateway device property that helps in deciding which gateway to choose.

Reply:
Ok, let's see how we fare with this approach ...